File: destack-py/destack/language/core/builtin/const.py
import contextvars
from datetime import datetime
from typing import (
    TYPE_CHECKING,
    Any,
    Optional,
    cast,
)

from destack.utils.env import get_from_env
from destack.utils.frozen import frozendict
from destack.utils.uuid import uuid4

from .common import Cloud, Region

if TYPE_CHECKING:
    from destack.language import Branch, Event, Node, NodeReference, Session, Snapshot, Space


class _Unset:
    def __repr__(self):
        return "<UNSET!>"

    def __str__(self):
        return "<UNSET!>"


# forever constants
VERSION = "2025.07.18.3"
FLOAT_EPSILON = 1e-6
BEGINNING_OF_TIME = datetime.fromisoformat("1970-01-01T00:00:00+00:00")


# runtime constants
NONCE = uuid4()
UNSET = cast(Any, _Unset())
EMPTY_LIST: list = []
EMPTY_SET: frozenset = frozenset()
EMPTY_DICT: dict[Any, Any] = frozendict()

# runtime context
ACTIVE_SESSION: contextvars.ContextVar[Optional["Session"]] = contextvars.ContextVar(
    "active_session", default=None
)
ACTIVE_SPACE: contextvars.ContextVar[Optional["Space"]] = contextvars.ContextVar(
    "active_space", default=None
)
ACTIVE_BRANCH: contextvars.ContextVar[Optional["Branch"]] = contextvars.ContextVar(
    "active_branch", default=None
)
ACTIVE_SNAPSHOT: contextvars.ContextVar[Optional["Snapshot"]] = contextvars.ContextVar(
    "active_snapshot", default=None
)
ACTIVE_EVENT: contextvars.ContextVar[Optional["Event[Node]"]] = contextvars.ContextVar(
    "active_event", default=None
)


def get_active_session() -> Optional["Session"]:
    """Gets the currently active Session (if any)."""
    return ACTIVE_SESSION.get()


def active_session() -> "Session":
    """Gets the currently active Session (error if none)."""
    session = ACTIVE_SESSION.get()
    assert session is not None, "no active session"
    return session


def get_active_space() -> Optional["Space"]:
    """Gets the currently active Space (if any)."""
    return ACTIVE_SPACE.get()


def get_active_space_ptr() -> Optional["NodeReference"]:
    """Gets the currently active Space (if any)."""
    space = ACTIVE_SPACE.get()
    return space.to_ref() if space else None


def active_space() -> "Space":
    """Gets the currently active Space (error if none)."""
    space = ACTIVE_SPACE.get()
    assert space is not None, "no active space"
    return space


def active_space_ptr() -> "NodeReference":
    """Gets the currently active Space (error if none)."""
    space = ACTIVE_SPACE.get()
    assert space is not None, "no active space"
    return space.to_ref()


def get_active_snapshot() -> Optional["Snapshot"]:
    """Gets the currently active Snapshot (if any)."""
    return ACTIVE_SNAPSHOT.get()


def get_active_branch() -> Optional["Branch"]:
    """Gets the currently active Branch (if any)."""
    return ACTIVE_BRANCH.get()


def active_branch() -> "Branch":
    """Gets the currently active Branch (error if none)."""
    branch = ACTIVE_BRANCH.get()
    assert branch is not None, "no active branch"
    return branch


def get_active_branch_ptr() -> Optional["NodeReference"]:
    """Gets the currently active Branch (if any)."""
    branch = ACTIVE_BRANCH.get()
    return branch.to_ref() if branch else None


def active_snapshot() -> "Snapshot":
    """Gets the currently active Snapshot (error if none)."""
    snapshot = ACTIVE_SNAPSHOT.get()
    assert snapshot is not None, "no active snapshot"
    return snapshot


def get_active_event() -> Optional["Event"]:
    """Gets the currently active Event (if any)."""
    return ACTIVE_EVENT.get()


def active_event() -> "Event":
    """Gets the currently active Event (error if none)."""
    event = ACTIVE_EVENT.get()
    assert event is not None, "no active event"
    return event


CLOUD = get_from_env("CLOUD", typ=Cloud, description="Cloud we're running in")
REGION = get_from_env("REGION", typ=Region, description="Region we're running in")
TRACING = get_from_env("TRACING", typ=bool, description="Enable tracing")


class DestackError(Exception):
    """Common base class for any regular errors."""

    pass

File: destack-ts/src/language/core/builtin/const.ts
import type { Branch, Session, Snapshot, Space } from "@destack/language";
import { uuid4 } from "@destack/utils";
import { ContextVar } from "@destack/utils/context";

// forever constants
export const VERSION = "2025.08.11.1";
export const EPSILON = 1e-6;
export const EPSILON_EXPONENT = 6;

// runtime constants
export const NONCE = uuid4();
export const UNSET = Symbol("UNSET");
export const EMPTY_LIST: any[] = [];
export const EMPTY_SET: Set<any> = new Set();
export const EMPTY_DICT: Record<string, any> = {};

// runtime context
export const ACTIVE_SESSION: ContextVar<Session | null> = new ContextVar(null);
export const ACTIVE_SPACE: ContextVar<Space | null> = new ContextVar(null);
export const ACTIVE_BRANCH: ContextVar<Branch | null> = new ContextVar(null);
export const ACTIVE_SNAPSHOT: ContextVar<Snapshot | null> = new ContextVar(null);
export const ACTIVE_EVENT: ContextVar<Event | null> = new ContextVar(null);

/** Gets the currently active Session (if any). */
export function getActiveSession(): Session | null {
  return ACTIVE_SESSION.get();
}

/** Gets the currently active Session (error if none). */
export function activeSession(): Session {
  const session = ACTIVE_SESSION.get();
  if (session == null) {
    throw new Error("no active Session");
  }
  return session;
}

/** Gets the currently active Space (if any). */
export function getActiveSpace(): Space | null {
  return ACTIVE_SPACE.get();
}

/** Gets the currently active Space (error if none). */
export function activeSpace(): Space {
  const space = ACTIVE_SPACE.get();
  if (space == null) {
    throw new Error("no active Space");
  }
  return space;
}

/** Gets the currently active Branch (if any). */
export function getActiveBranch(): Branch | null {
  return ACTIVE_BRANCH.get();
}

/** Gets the currently active Branch (error if none). */
export function activeBranch(): Branch {
  const branch = ACTIVE_BRANCH.get();
  if (branch == null) {
    throw new Error("no active Branch");
  }
  return branch;
}

/** Gets the currently active Snapshot (if any). */
export function getActiveSnapshot(): Snapshot | null {
  return ACTIVE_SNAPSHOT.get();
}

/** Gets the currently active Snapshot (error if none). */
export function activeSnapshot(): Snapshot {
  const snapshot = ACTIVE_SNAPSHOT.get();
  if (snapshot == null) {
    throw new Error("no active Snapshot");
  }
  return snapshot;
}

/** Gets the currently active Event (if any). */
export function getActiveEvent(): Event | null {
  return ACTIVE_EVENT.get();
}

/** Gets the currently active Event (error if none). */
export function activeEvent(): Event {
  const event = ACTIVE_EVENT.get();
  if (event == null) {
    throw new Error("no active Event");
  }
  return event;
}
